dothething/dothething/ClipsView.swift

```diff
@@ -35,71 +34,101 @@ struct ClipsView: View {
-                }) {
-                    Text("Share")
+                
+                ScrollView {
+                    LazyVGrid(columns: [GridItem](repeating: GridItem(.flexible(), spacing: 4), count: 3), spacing: 4) {
+                        ForEach(clipsViewModel.clips, id: \.self) { clip in
+                            ClipView(clip: clip)
+                        }
+                        .frame(height: (192 / 108) * geometry.size.width / 3)
+                    }
+                    .padding(.leading, 4)
+                    .padding(.trailing, 4)
                 }
-                .padding(.trailing)
-                .frame(maxWidth: .infinity, alignment: .trailing)
-                .opacity(clipsViewModel.shareDisabled ? 0.5 : 1)
-            }
-
-            if !clipsViewModel.errorText.isEmpty && clipsViewModel.clips.isEmpty {
-                Text(clipsViewModel.errorText)
-                    .padding(.top)
-                    .foregroundColor(Color(UIColor.systemGray))
-            } else if clipsViewModel.isLoading {
-                ProgressView()
-                    .progressViewStyle(.circular)
-                    .padding(.top)
-            }
-            
-            ScrollView {
-                LazyVGrid(columns: [GridItem(.adaptive(minimum: 108, maximum: 108), spacing: 16)]) {
-                    ForEach(clipsViewModel.clips, id: \.self) { clip in
-                        ClipView(clip: clip)
+                .padding(.top, -12)
+                .padding(.bottom, -12)
+                
+                ZStack {
+                     Rectangle()
+                        .frame(height: 48)
+                        .foregroundColor(.accentColor)
+
+                    Text("code: \(clipsViewModel.code)")
+                        .font(.custom("Montserrat-Light", size: 24))
+                        .foregroundColor(.accentColor)
+                        .colorInvert()
+                        .onTapGesture {
+                            print("Code tapped")
+                            UIPasteboard.general.string = clipsViewModel.code
+                        }
+                }
+                
+                Button(action: {
+                    if clipsViewModel.buttonText == "place a domino" {
+                        clipsViewModel.uploadButtonPressed()
+                    } else if clipsViewModel.buttonText == "share this rally" {
+                        clipsViewModel.shareButtonPressed()
                     }
+                }) {
+                    Text(clipsViewModel.buttonText)
+                        .font(.custom("Montserrat-Light", size: 24, relativeTo: .title))
+                        .foregroundColor(.accentColor)
+                        .colorInvert()
+                        .padding()
+                        .frame(width: 300, height: 50)
+                        .background(Color.accentColor)
+                        .cornerRadius(50)
```

Comment:
The new code adds a "place a domino" button at the bottom of the view, which is a more logical placement for this action. The button also has a nice visual design, with a custom font and background color. The code also includes a "share this rally" button, which is a relevant addition to the functionality of the app.